### PR TITLE
Introduce XMPP pipelining.

### PR DESCRIPTION
**Pull request by kzemek**
This PR introduces pipelining via updating exml to a version that can self-reset XMPP streams, allowing to parse payloads with <stream:stream> tag and XML declaration in the middle.

### CODE REVIEW

File: apps/ejabberd/src/ejabberd_s2s_in.erl [hunk 1]
```diff
@@ -360,8 +360,6 @@ wait_for_feature_request({xmlstreamelement, El}, StateData) ->
                         end,
                     if
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_s2s_in.erl]: According to **Elvis**:
> Replace the 'if' expression on line 361 with a 'case' expression or function clauses.

File: apps/ejabberd/src/mod_websockets.erl [hunk 2]
```diff
@@ -317,7 +303,7 @@ replace_stream_ns(Element, _State) ->
     Element.
 
 get_parser_opts(<<"<open", _/binary>>) -> [{infinite_stream, true}, {autoreset, true}]; % new-type WS
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_websockets.erl]: According to **Elvis**:
> Line 305 is too long: get_parser_opts(<<"<open", _/binary>>) -> [{infinite_stream, true}, {autoreset, true}]; % new-type WS.

File: apps/ejabberd/src/mod_websockets.erl [hunk 3]
```diff
@@ -112,11 +111,6 @@ websocket_info({send_xml, XML}, Req, State) ->
     XML1 = process_server_stream_root(replace_stream_ns(XML, State), State),
     Text = exml:to_iolist(XML1),
     {reply, {text, Text}, Req, State};
-websocket_info(reset_stream, Req, #ws_state{parser = undefined} = State) ->
-    {ok, Req, State};
-websocket_info(reset_stream, Req, #ws_state{parser = Parser} = State) ->
-    {ok, NewParser} = exml_stream:reset_parser(Parser),
-    {ok, Req, State#ws_state{ parser = NewParser, open_tag = undefined }};
 websocket_info({set_ping, Value}, Req, State = #ws_state{ping_rate = none}) when is_integer(Value) and (Value > 0)->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_websockets.erl]: According to **Elvis**:
> Line 114 is too long: websocket_info({set_ping, Value}, Req, State = #ws_state{ping_rate = none}) when is_integer(Value) and (Value > 0)->.

File: apps/ejabberd/src/ejabberd_s2s_in.erl (hunk 1 above)
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_s2s_in.erl]: According to **Elvis**:
> The expression on line 361 and column 21 is nested beyond the maximum level of 3.

File: rebar.config [hunk 4]
```diff
@@ -9,7 +9,7 @@
   {base16, ".*", {git, "git://github.com/goj/base16.git", "f78918e"}},
   {cuesport, ".*", {git, "git://github.com/esl/cuesport.git", "d82ff25"}},
   {redo, ".*", {git, "git://github.com/Wallapop/redo.git", "35a8d1c"}},
-  {exml, ".*", {git, "git://github.com/esl/exml.git", "2.4.1"}},
+  {exml, ".*", {git, "git://github.com/esl/exml.git", {ref, "7b5bfca"}}},
```
Comment (michalwski) [inline rebar.config]: Could you change exml ref to https://github.com/esl/exml/commit/d365533c193fe61fd211c74518b2136ef71b60ec ?

File: test.disabled/ejabberd_tests/rebar.config [hunk 5]
```diff
@@ -4,16 +4,14 @@
...
         {erlsh, ".*", {git, "git://github.com/proger/erlsh.git", "2fce513"}},
         {jiffy, ".*", {git, "git://github.com/davisp/jiffy.git", "0.14.8"}},
         {proper, ".*", {git, "https://github.com/manopapad/proper.git", "v1.2"}},
         {cowboy, ".*", {git, "git://github.com/ninenines/cowboy.git", {tag, "1.0.4"}}},
-        {shotgun, ".*", {git, "https://github.com/inaka/shotgun.git", "4e67065"}}
-
-
+        {shotgun, ".*", {git, "https://github.com/inaka/shotgun.git", "4e67065"}},
+        {exml, ".*", {git, "git://github.com/esl/exml.git", "7b5bfca"}},
+        {escalus, ".*", {git, "git://github.com/esl/escalus.git", "8f0e101"}}
```
Comment (michalwski) [inline test.disabled/ejabberd_tests/rebar.config]: Looks like we need a PR in [esl/escalus](https://github.com/esl/escalus)

**Comment by michalwski** [comment]
Thanks!

**Pull request merged**